frontend/src/app/shared/validation/join-id-or-url-validator.directive.ts

```diff
@@ -0,0 +1,20 @@
+import { Directive } from '@angular/core';
+import type { AbstractControl, Validator } from '@angular/forms';
+import { NG_VALIDATORS } from '@angular/forms';
+import { CustomValidators } from './custom-validators';
+
+@Directive({
+    selector: '[appJoinUrlOrIdValidator]',
+    providers: [
+        {
+            provide: NG_VALIDATORS,
+            useExisting: JoinIdOrUrlDirective,
+            multi: true,
+        },
+    ],
+})
+export class JoinIdOrUrlDirective implements Validator {
+    validate(control: AbstractControl) {
+        return CustomValidators.joinUrlOrIdValidator()(control);
```

Comment:
All validators could probably use some refactoring to not recreate the validator from the factory on every key press...

Reply:
Since I'm not that familiar with Angular, I'm not sure what is the best way to solve this.

My idea was using the factory in the constructor and storing the validator function, so the code would look like this:

```ts
export class JoinIdDirective implements Validator {
    private readonly validator: (control: AbstractControl) => any;
    constructor() {
        this.validator = CustomValidators.joinIdValidator();
    }
    validate(control: AbstractControl) {
        return this.validator(control);
    }
}
```

This would still create one validator per input element, but at least not per key press. I think it should also work for validators with inputs (e.g. `exactMatchValidator`) by using `implements OnChanges`/`ngOnChanges` to (re-)create the validator.

If you think this is okay, I would change the `joinIdValidator` as stated above and refactor the other validators in a separate PR

Reply:
You could simplify this a bit:

```ts
export class JoinIdDirective implements Validator {
    private readonly validator = CustomValidators.joinIdValidator();

    validate(control: AbstractControl) {
        return this.validator(control);
    }
}
```

For the factories with parameters, `OnChanges` should work:

```ts
export class ExactMatchValidatorDirective implements Validator, OnChanges {
    /**
     * The string to match
     */
    @Input() appExactMatchValidator!: string;

    private validator!: ValidatorFn;

    validate(control: AbstractControl) {
        return this.validator(control);
    }

    ngOnChanges(changes: SimpleChangesGeneric<this>) {
        this.validator = CustomValidators.exactMatchValidator(
            this.appExactMatchValidator
        );
    }
}
```

One could even think about changing the `CustomValidators` from only Validator-factory-functions to Validators or Validator-factory-functions. But the performance benefit would be negligible, and the typings would become more complicated.

> If you think this is okay, I would change the joinIdValidator as stated above and refactor the other validators in a separate PR

👍

Reply:
Created #559 for this. Will start working on it after this PR is merged to prevent conflicts due to the addition and deletion of validators.

The `appJoinIdValidator` is refactored now.